src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs

```diff
@@ -206,115 +218,162 @@ protected NodeContext GetNode(string msbuildLocation, string commandLineArgs, in
                 }
             }
 
+            // Get all process of possible running node processes for reuse and put them into ConcurrentQueue.
+            // Processes from this queue will be concurrently consumed by TryReusePossibleRunningNodes while
+            //    trying to connect to them and reuse them. When queue is empty, no process to reuse left
+            //    new node process will be started.
+            string expectedProcessName = null;
+            ConcurrentQueue<Process> possibleRunningNodes = null;
```

Comment:
Is there any reason you're using ConcurrentQueue as opposed to some other concurrent data structure? It has tighter constraints that ConcurrentBag, so I'd imagine that's faster? (Haven't tested that.)

Reply:
I thought about it, I liked Q better as it preserver order of creation which could help with troubleshooting. Since there is no `Concurrent ordered collection` I am using `ConcurrentQueue` whenever I need ordered concurrent collection.
In this context when we are spending 100ms spawning process the difference between queue and bag would be insignificant. I have not measured it yet but I expect it will be < 1ns.

Reply:
I'm thinking about the limit as number of processors approaches infinity. It should get progressively harder to use the queue concurrently, which isn't true for a bag. No computer actually has infinite processors of course, but people doing C++ seem to have a lot. Might be reasonable?

Reply:
When in doubt I call DotNetBenchmark for a help. Surprisingly, ConcurrentQueue is faster that ConcurrentBag.
![image](https://user-images.githubusercontent.com/25249058/163484185-8873f21a-94c3-4b93-8e9f-1b60f2f0b165.png)
Result of my benchmark is consitent with this article: https://minidump.net/performance-best-practices-in-c-b85a47bdd93a

Reply:
Also please note that for 1000 items (1000 processes which match our process name) it is ~1ns while connecting to ONE node is 20ms which  is 0.005% :-)

Reply:
What code did you use for your benchmark? I was partially surprised because I'd run my own benchmark and got very different results:
| Method | NumCases |         Mean |      Error |       StdDev |       Median |     Gen 0 |     Gen 1 |     Gen 2 |   Allocated |
|------- |--------- |-------------:|-----------:|-------------:|-------------:|----------:|----------:|----------:|------------:|
|    Bag |      100 |     17.26 us |   0.336 us |     0.588 us |     17.14 us |    0.7629 |    0.3662 |         - |     4.67 KB |
|  Queue |      100 |     13.44 us |   0.150 us |     0.133 us |     13.43 us |    0.8545 |         - |         - |     5.23 KB |
|    Bag |    10000 |    262.59 us |   8.581 us |    24.758 us |    253.76 us |   31.7383 |   15.6250 |    2.9297 |   155.39 KB |
|  Queue |    10000 |    692.55 us |  13.848 us |    39.510 us |    679.86 us |   23.4375 |    5.8594 |         - |   134.14 KB |
|    Bag |  1000000 | 12,165.84 us | 335.746 us |   989.955 us | 12,124.86 us | 1359.3750 | 1281.2500 | 1046.8750 | 16686.98 KB |
|  Queue |  1000000 | 40,737.60 us | 808.214 us | 1,857.007 us | 40,433.64 us |  666.6667 |  666.6667 |  666.6667 |   8215.2 KB |

The article mentioned that ConcurrentBag is optimized for the process that uses something (normally) being the same as the one that added it to the bag. My code essentially looked like this:
```C#
ConcurrentContainer<int> container = new();
Parallel.For(0, NumCases, i =>
{
    container.Add(i);
});
foreach (int i in container);
```

Reply:
Most noteworthy, I think, is that for larger numbers of threads adding, the Bag starts to be much faster than the queue, but its allocations also get to be much larger.

Reply:
(To be clear, I think it's reasonable/fine to use a queue; I'm just curious why your numbers would be so unexpected.)

Reply:
But this is not how we use it in this code. Your benchmark have concurrent producers and one thread consumer. Our usage is one thread producer - get processes -  and multi thread competing consumers.
Try to measure it with something like this:
```
ConcurrentContainer<int> container = new();
for (int i = 0; i < NumCases; i++)
   container.Add(i);
Parallel.For(0, NumCases, _ =>
{
    container.TryTake(out _);
});
```

Reply:
I added one little tweak to your version: rather than initializing a concurrent collection then adding to it, I initialized a normal List, added to it, then used the concurrent constructor taking a list.

| Method | NumCases |         Mean |        Error |       StdDev |     Gen 0 |     Gen 1 |     Gen 2 |   Allocated |
|------- |--------- |-------------:|-------------:|-------------:|----------:|----------:|----------:|------------:|
|    Bag |      100 |     23.07 us |     0.455 us |     0.447 us |    0.8545 |    0.4272 |         - |     5.29 KB |
|  Queue |      100 |     13.97 us |     0.278 us |     0.285 us |    0.8240 |         - |         - |     5.02 KB |
|    Bag |    10000 |  1,467.20 us |    41.087 us |   119.854 us |   42.9688 |   21.4844 |    3.9063 |   266.05 KB |
|  Queue |    10000 |    966.75 us |    26.234 us |    77.350 us |   41.0156 |   41.0156 |   41.0156 |   260.71 KB |
|    Bag |  1000000 | 86,002.80 us | 1,717.738 us | 4,403.219 us | 1800.0000 | 1800.0000 | 1800.0000 |  16397.6 KB |
|  Queue |  1000000 | 81,010.31 us | 2,224.135 us | 6,200.000 us |  625.0000 |  625.0000 |  625.0000 | 16398.53 KB |

What I think is most interesting here is that Queue trounces Bag for small input sizes, but they approach the same for larger inputs. I'd wager that if I went for even NumCases == 10,000,000, Bag would be strictly more performant than Queue. That said, if someone has millions of MSBuild processes running on one computer all at once, I think we have a more fundamental problem...